File: src/common/feature_flags.py
"""Feature-flag evaluation backed by OpenFeature.

Vendored into every function's deployment zip. Flag state lives in a single **AWS
SSM Parameter Store** parameter (a standard-tier ``String`` holding the flag JSON,
one per environment) and is read at runtime through the boto3 ``ssm``
``GetParameter`` API — so toggling a flag is an edit to the parameter value in the
SSM console, with **no redeploy**. Standard-tier parameters are free for both
storage and ``GetParameter`` calls. See
``docs/requirements/backend/BE-017-feature-flags.md``.

SSM is selected only when ``FEATURE_FLAGS_SSM_PARAM`` is set (the deployed
Lambdas). Otherwise — local dev and tests — there is **no flag source** and every
flag defaults to ``False`` (feature off). There is no bundled JSON fallback.

A ``billing`` master flag gates all Stripe billing behavior (BE-014 / BE-015):
when it is OFF, ``require_active_subscription`` is a no-op, the checkout and
billing-portal endpoints return 404, and the Stripe webhook no-ops.

On top of it, the ``premium_feature`` flag implements the freemium model: a
premium feature is paywalled only when **both** ``billing`` and ``premium_feature``
are ON (see ``is_feature_paywalled``). Every premium feature shares this one flag,
so they are all gated identically. The frontend gates the schedule-swap simulator
(FE-031) on it; no backend endpoint enforces it yet.

Evaluation goes through OpenFeature's in-memory provider so the rest of the app
depends only on the vendor-neutral OpenFeature client. Anything that cannot be
resolved — SSM unreachable, an unknown flag — fails safe to ``False``
(feature off).
"""

import json
import logging
import os
import time

import boto3
from openfeature import api
from openfeature.provider.in_memory_provider import InMemoryFlag, InMemoryProvider

logger = logging.getLogger(__name__)

# Shared premium-feature flag (freemium model; see BE-014 / BE-017). Every premium
# feature is gated identically on this one flag. The frontend gates the schedule-swap
# simulator (FE-031) on it; no backend endpoint enforces it yet.
PREMIUM_FEATURE = "premium_feature"

# Global, non-billing flag gating the in-app informational banner (FE-030) — a
# generic toggle reused for whatever the current banner promotes (Discord today).
# Surfaced to the SPA via GET /feature-flags; the backend enforces nothing.
BANNER = "banner"

# The variant names are cosmetic; what matters is the boolean each maps to.
_ON = "on"
_OFF = "off"

# SSM wiring (BE-017). When the parameter name is present the flag JSON is sourced
# from AWS SSM Parameter Store; otherwise (local / tests) every flag defaults off.
_FEATURE_FLAGS_SSM_PARAM = os.environ.get("FEATURE_FLAGS_SSM_PARAM")
# How long a fetched configuration is reused before the next poll (seconds). A
# console edit is picked up within this window.
_FEATURE_FLAGS_TTL_SECONDS = int(os.environ.get("FEATURE_FLAGS_TTL_SECONDS", "45"))

_flags_enabled = bool(_FEATURE_FLAGS_SSM_PARAM)
# Module-level boto3 client, mirroring the codebase's ssm/dynamo pattern. Only
# created when SSM is wired, so tests never touch the network.
_ssm_client = boto3.client("ssm") if _flags_enabled else None

# Mutable TTL cache state.
_cached_config: dict = {}
_last_refresh_at: float = 0.0


def _build_flags(config: dict) -> dict[str, InMemoryFlag]:
    """Turn a ``{name: {"enabled": bool}}`` config into OpenFeature flags."""
    flags: dict[str, InMemoryFlag] = {}
    for name, spec in config.items():
        enabled = bool(spec.get("enabled", False)) if isinstance(spec, dict) else False
        flags[name] = InMemoryFlag(
            default_variant=_ON if enabled else _OFF,
            variants={_ON: True, _OFF: False},
        )
    return flags


def _set_provider_from_config(config: dict) -> None:
    api.set_provider(InMemoryProvider(_build_flags(config)))


def _fetch_flags() -> dict:
    """Pull the latest flag JSON from the SSM parameter.

    The feature-flag parameter holds the same ``{name: {"enabled": bool}}`` shape
    ``_build_flags`` already parses. A missing parameter raises ``ParameterNotFound``,
    handled by the caller as a fetch error (fail-safe to the last-known / all-off
    config).
    """
    response = _ssm_client.get_parameter(Name=_FEATURE_FLAGS_SSM_PARAM)
    raw = response["Parameter"]["Value"]
    if not raw:
        return {}
    return json.loads(raw)


def _refresh_if_stale() -> None:
    """Refresh the cached flags from SSM once the TTL has elapsed.

    A fetch failure (network, missing parameter, malformed JSON) keeps the
    last-known flags — flags never flip to a surprise state because SSM hiccuped.
    """
    global _cached_config, _last_refresh_at
    now = time.monotonic()
    if now - _last_refresh_at < _FEATURE_FLAGS_TTL_SECONDS:
        return
    _last_refresh_at = now
    try:
        config = _fetch_flags()
    except Exception as exc:
        logger.warning("Feature-flag refresh failed (%s); using last-known flags", exc)
        return
    if config != _cached_config:
        _cached_config = config
        _set_provider_from_config(config)


def _initialize() -> None:
    """Seed the provider at import: from SSM when wired, else all-off."""
    global _cached_config, _last_refresh_at
    if not _flags_enabled:
        _set_provider_from_config({})
        return
    try:
        _cached_config = _fetch_flags()
    except Exception as exc:
        logger.warning(
            "Initial feature-flag fetch failed (%s); all flags default off", exc
        )
        _cached_config = {}
    _last_refresh_at = time.monotonic()
    _set_provider_from_config(_cached_config)


_initialize()
_client = api.get_client()


def is_enabled(flag_name: str) -> bool:
    """Return whether ``flag_name`` is on, defaulting to ``False`` when unknown."""
    if _flags_enabled:
        _refresh_if_stale()
    return _client.get_boolean_value(flag_name, False)


def is_billing_enabled() -> bool:
    """Return whether Stripe billing (BE-014 / BE-015) is enabled."""
    return is_enabled("billing")


def is_feature_paywalled(flag_name: str) -> bool:
    """Return whether a premium feature is paywalled (freemium model; BE-014).

    A feature is paywalled only when **both** the ``billing`` master flag and the
    feature's own ``flag_name`` are ON. With billing off (the default) this is
    always ``False``, so every feature is free.
    """
    return is_billing_enabled() and is_enabled(flag_name)


def _override_for_testing(flags: dict[str, bool]) -> None:
    """Replace the active provider with an explicit flag map (tests only).

    Lets a test exercise the ON path without standing up SSM:
    ``_override_for_testing({"billing": True})``.
    """
    _set_provider_from_config({name: {"enabled": on} for name, on in flags.items()})

File: frontend/src/lib/feature-flags.ts
/**
 * Feature-flag evaluation backed by OpenFeature (FE-026).
 *
 * Flag state lives in **AWS SSM Parameter Store** (global, per environment) and is
 * resolved at runtime from the backend's public `GET /feature-flags` endpoint — so a
 * console toggle reaches the SPA without a rebuild. There is no bundled config:
 * until {@link initFeatureFlags} resolves (and any time the backend is
 * unreachable) every flag fails safe to `false` (feature off).
 *
 * A `billing` flag currently gates all subscription UI (FE-021/022/023): when it
 * is off, `SubscriptionGuard` is a pass-through and the "Manage Subscription"
 * sidebar entry is hidden.
 *
 * Evaluation goes through OpenFeature's in-memory provider so the rest of the app
 * depends only on the vendor-neutral OpenFeature client. An unknown flag fails
 * safe to `false`.
 */
import { OpenFeature, InMemoryProvider } from '@openfeature/web-sdk';

import { API_BASE_URL } from '@/lib/api-client';

interface FlagSpec {
  enabled?: boolean;
}

/** Convert the `{ name: { enabled } }` config into an OpenFeature flag map. */
function toFlagConfiguration(config: Record<string, FlagSpec>) {
  return Object.fromEntries(
    Object.entries(config).map(([name, spec]) => [
      name,
      {
        variants: { on: true, off: false },
        defaultVariant: spec?.enabled ? 'on' : 'off',
        disabled: false,
      },
    ]),
  );
}

/**
 * Canonical serialization of a flag map (name → enabled, key-sorted) used to
 * detect whether the resolved flags actually changed. A refresh that returns the
 * same values — e.g. the `visibilitychange` refresh on every tab focus — must NOT
 * swap the provider, because swapping emits `PROVIDER_READY` /
 * `PROVIDER_CONFIGURATION_CHANGED`, which remounts the whole app (FE-026).
 */
function serializeFlags(config: Record<string, FlagSpec>): string {
  return JSON.stringify(
    Object.entries(config)
      .map(([name, spec]) => [name, spec?.enabled ?? false] as const)
      .sort(([a], [b]) => a.localeCompare(b)),
  );
}

/** The flag map currently installed on the active provider (see `serializeFlags`). */
let lastAppliedFlags = '';

/** Build the `InMemoryProvider` for a flag config and register it globally. */
function setProviderFromConfig(config: Record<string, FlagSpec>): void {
  lastAppliedFlags = serializeFlags(config);
  OpenFeature.setProvider(new InMemoryProvider(toFlagConfiguration(config)));
}

// Until initFeatureFlags() resolves the real values from the backend, every flag
// is off — there is no bundled config to fall back on. Also the fail-safe state
// whenever the flags endpoint is unreachable.
setProviderFromConfig({});

const client = OpenFeature.getClient();

/** Whether `flagName` is on; unknown flags default to `false`. */
export function isEnabled(flagName: string): boolean {
  return client.getBooleanValue(flagName, false);
}

/** Whether subscription/billing UI (FE-021/022/023) is enabled. */
export function isBillingEnabled(): boolean {
  return isEnabled('billing');
}

/** Whether the in-app informational banner (FE-030) is enabled. */
export function isBannerEnabled(): boolean {
  return isEnabled('banner');
}

/** Vitest sets this sentinel (see vite.config.ts); flag fetching stays off in tests. */
function isTestEnv(): boolean {
  return import.meta.env.VITE_API_URL === 'http://test.local';
}

/** Shape of the public `GET /feature-flags` payload (`{ name: enabled }` under `data`). */
interface FeatureFlagsPayload {
  data?: Record<string, boolean>;
}

/**
 * Fetch the global flags and register them. Any failure (non-200, network error,
 * unreachable backend) leaves the current provider in place, so the app keeps the
 * last-known — or the fail-safe all-off — flags rather than throwing.
 *
 * Exported for tests (the mapping seam); production code calls it via
 * {@link initFeatureFlags}, which gates it behind the Vitest check.
 */
export async function refreshFlags(): Promise<void> {
  try {
    const res = await fetch(`${API_BASE_URL}/feature-flags`);
    if (!res.ok) return;
    const body = (await res.json()) as FeatureFlagsPayload;
    const flags = body.data ?? {};
    const config = Object.fromEntries(
      Object.entries(flags).map(([name, enabled]) => [name, { enabled }]),
    );
    // Only swap the provider when the values actually changed. An unchanged
    // refresh (the common case for the visibilitychange/poll refresh) must be a
    // no-op so it does not emit a provider event and remount the app (FE-026).
    if (serializeFlags(config) === lastAppliedFlags) return;
    setProviderFromConfig(config);
  } catch {
    // Keep the fail-safe provider; a transient outage must not flip flags on.
  }
}

// How often the SPA re-polls the flags so a console toggle is picked up without a
// reload. Paired with a refresh whenever the tab regains focus.
const REFRESH_INTERVAL_MS = 60_000;

let refreshStarted = false;

/**
 * Resolve feature flags from the backend and keep them fresh. Called once at
 * bootstrap (see `app/main.tsx`) before first render. A no-op under Vitest so
 * component tests never hit the network (MSW runs `onUnhandledRequest: 'error'`);
 * those drive flags via {@link setFlagsForTesting} instead.
 */
export async function initFeatureFlags(): Promise<void> {
  if (isTestEnv()) return;
  await refreshFlags();
  if (refreshStarted) return;
  refreshStarted = true;
  setInterval(() => void refreshFlags(), REFRESH_INTERVAL_MS);
  document.addEventListener('visibilitychange', () => {
    if (document.visibilityState === 'visible') void refreshFlags();
  });
}

/**
 * Replace the active provider with an explicit flag map (tests only). Lets a test
 * exercise the billing-on path without standing up the backend flags endpoint.
 */
export function setFlagsForTesting(flags: Record<string, boolean>): void {
  setProviderFromConfig(
    Object.fromEntries(
      Object.entries(flags).map(([name, enabled]) => [name, { enabled }]),
    ),
  );
}
